### PR TITLE
Add GitHub Actions workflow

### PR DESCRIPTION
**Pull request by athomas**

### CODE REVIEW

**Comment by athomas** [comment]
Not sure if the workflow is going to trigger here, but I had it running in my fork:
https://github.com/athomas/pub/actions/runs/423098796

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        shard: [0, 1, 2, 3, 4, 5, 6]
+    steps:
+      - uses: actions/checkout@v2
+      - name: Cache Pub hosted dependencies
+        uses: actions/cache@v2
+        with:
+          path: "~/.pub-cache/hosted"
+          key: "os:ubuntu-latest;pub-cache-hosted;dart:${{ matrix.sdk }};packages:pub"
+          restore-keys: |
+            os:${{ matrix.os }};pub-cache-hosted;dart:${{ matrix.sdk }};packages:pub
```
Comment (sigurdm) [inline .github/workflows/test.yaml:66]: I don't really understand these lines.
Could you add a comment what they do?

**Review by sigurdm — APPROVED** [review]
Basically looks good to me. Just a few suggestions

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+        sdk: [dev]
+        shard: [0, 1, 2, 3, 4, 5, 6]
+    steps:
+      - uses: actions/checkout@v2
+      - name: Cache Pub hosted dependencies
+        uses: actions/cache@v2
+        with:
+          path: "~/.pub-cache/hosted"
```
[the remark this answers is not in the document]
Comment (athomas) [reply .github/workflows/test.yaml]: Thanks, after looking into the caching some more I've decided to remove the caches.

File: .github/workflows/test.yaml (hunk 1 above)
Comment (athomas) [reply .github/workflows/test.yaml]: They are a prefix search for less specific cache keys, it makes some sense in mono repos but not really here. Anyway, I deleted the whole cache step.

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -0,0 +1,86 @@
+name: Test
```
[the remark this answers is not in the document]
Comment (athomas) [reply .github/workflows/test.yaml:1]: Done.

File: .github/workflows/test.yaml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        with:
+          channel: ${{ matrix.sdk }}
+      - name: Install dependencies
+        run: dart pub get
+      - name: Build snapshot
+        run: dart --snapshot=${{ env._PUB_TEST_SNAPSHOT }} bin/pub.dart
+      - name: Select tests
+        run: |
+          find test -name "*_test\\.dart" | sort > $_TESTS_FILE
+          (( tests_per_shard = ($(wc -l <$_TESTS_FILE) + 7 - 1) / 7 ))
```
[the remark this answers is not in the document]
Comment (athomas) [reply .github/workflows/test.yaml:60]: The `split` solution is tricky on macOS where the split version is really old. On travis, the runner was installed with coreutils from homebrew which made a modern `split` available as `gsplit`. GitHub Actions doesn't have that kind of install step, so we'd have to complicate the workflow a fair amount to get the same result. That's why I went with the bash magic.

**Comment by sigurdm** [comment]
I think you can add a skip+TODO for the failing tests, and let us investigate later

**Comment by athomas** [comment]
> I think you can add a skip+TODO for the failing tests, and let us investigate later

`nslookup pub.invalid` results in a `SERVFAIL` on Azure instead of an `NXDOMAIN` error. Changing the url to something else (`invalid-url.foo` which was used in a similar `pub add` test) fixed the broken test.

**Pull request merged**